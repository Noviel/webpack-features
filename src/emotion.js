export default ({ target, production }, { plugin = true } = {}, state) => {
  if (plugin) {
    const babelLoaders = state.loaders(
      ({ loader }) => loader === 'babel-loader'
    );

    for (const loader of babelLoaders) {
      loader.options.plugins.unshift([
        'emotion',
        { extractStatic: target.browsers && production },
      ]);
    }
  }
};
