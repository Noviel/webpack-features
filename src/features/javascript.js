import createBabelEnvPresetOptions from '../lib/babelEnvPresetOptions';

const syntaxExtendPlugins = [
  '@babel/plugin-proposal-object-rest-spread',
  '@babel/plugin-proposal-decorators',
  ['@babel/plugin-proposal-class-properties', { loose: true }],
  '@babel/plugin-syntax-dynamic-import',
];

const createPresetsList = ({ env, react, flow }) =>
  []
    .concat(
      env ? [['@babel/preset-env', createBabelEnvPresetOptions(env)]] : []
    )
    .concat(react ? '@babel/preset-react' : [])
    .concat(flow ? '@babel/preset-flow' : []);

export default (
  env,
  {
    eslint = true,
    flow = true,
    modules = false,
    react = true,
    syntaxExtend = true,
    hot = !env.production && env.target.name === 'browsers',
    babelPlugins = [],
    exclude = /node_modules/,
  } = {},
  { plugins = [], next }
) => {
  const { target } = env;
  const rules = [
    {
      test: /\.jsx?$/,
      exclude,

      use: []
        .concat({
          loader: 'babel-loader',
          options: {
            babelrc: false,
            presets: createPresetsList({
              env: { target, modules },
              react,
              flow,
            }),
            plugins: []
              .concat(hot && react ? 'react-hot-loader/babel' : [])
              .concat(syntaxExtend ? syntaxExtendPlugins : [])
              .concat(babelPlugins)
              .concat(hot && react ? '@babel/transform-classes' : [])
              .concat(hot && react ? '@babel/transform-react-jsx-source' : []),
          },
        })
        .concat(
          eslint
            ? {
                loader: 'eslint-loader',
                options: typeof eslint === 'object' ? eslint : {},
              }
            : []
        ),
    },
  ];

  return next(env, plugins, { module: { rules } });
};
