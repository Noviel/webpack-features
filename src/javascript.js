import createBabelEnvPresetOptions from './babel-env-preset-options';

const syntaxExtendPlugins = [
  'transform-object-rest-spread',
  'transform-class-properties',
  'syntax-dynamic-import',
];

const createPresetsList = ({ env, react, flow }) =>
  []
    .concat(env ? [['env', createBabelEnvPresetOptions(env)]] : [])
    .concat(react ? 'react' : [])
    .concat(flow ? 'flow' : []);

const createEslintLoader = options => ({
  loader: 'eslint-loader',
  options: typeof options === 'object' ? options : {},
});

const createBabelLoader = (
  { target, production },
  { modules, react, flow, plugins, syntaxExtend, hot }
) => ({
  loader: 'babel-loader',
  options: {
    babelrc: false,
    presets: createPresetsList({ env: { ...target, modules }, react, flow }),
    plugins: []
      .concat(hot && react ? 'react-hot-loader/babel' : [])
      .concat(syntaxExtend ? syntaxExtendPlugins : [])
      .concat(plugins)
      .concat(hot && react ? 'transform-es2015-classes' : [])
      .concat(hot && react ? 'transform-react-jsx-source' : []),
  },
});

module.exports = (
  { target, production },
  {
    eslint = true,
    flow = true,
    modules = false,
    react = true,
    syntaxExtend = true,
    hot = !production && !!target.browsers,
    plugins = [],
  } = {},
  state
) => {
  state.addRules({
    test: /\.jsx?$/,
    exclude: /node_modules/,

    use: []
      .concat(
        createBabelLoader(
          {
            target,
            production,
          },
          { modules, react, flow, plugins, syntaxExtend, hot }
        )
      )
      .concat(eslint ? createEslintLoader(eslint) : []),
  });
};
