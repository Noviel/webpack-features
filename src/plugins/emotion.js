export default ({ extractStatic = false } = {}) => (
  { target, production },
  result
) => {
  const babelLoaders = result.module.rules[0].use.filter(
    ({ loader }) => loader === 'babel-loader'
  );

  for (const loader of babelLoaders) {
    loader.options.plugins.unshift([
      'emotion',
      {
        extractStatic,
      },
    ]);
  }

  return result;
};
