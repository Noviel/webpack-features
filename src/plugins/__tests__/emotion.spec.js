import emotion from '../emotion';
import javascript from '../../features/javascript';
import { callFeature } from '../../fixtures';

describe(`emotion`, () => {
  it(`should add emotion to the plugins list of the babel-loader's options`, () => {
    const result = callFeature(javascript, {}, [emotion()]).module.rules[0]
      .use[0].options.plugins;
    expect(result).toMatchSnapshot();
  });
});
