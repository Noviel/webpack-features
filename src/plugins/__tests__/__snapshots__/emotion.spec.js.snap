// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emotion should add emotion to the plugins list of the babel-loader's options 1`] = `
Array [
  Array [
    "emotion",
    Object {
      "autoLabel": false,
      "extractStatic": false,
      "hoist": true,
      "sourceMap": false,
    },
  ],
  "@babel/plugin-proposal-object-rest-spread",
  "@babel/plugin-proposal-decorators",
  Array [
    "@babel/plugin-proposal-class-properties",
    Object {
      "loose": true,
    },
  ],
  "@babel/plugin-syntax-dynamic-import",
]
`;
